src/python/pants/engine/rules.py

```diff
@@ -343,18 +385,38 @@ def wrapper(*args):
         return wrapper
 
 
+F = TypeVar("F", bound=Callable[..., Any | Coroutine[Any, Any, Any]])
+
+
 @overload
-def rule(func: Callable[P, Coroutine[Any, Any, R]]) -> Callable[P, Coroutine[Any, Any, R]]: ...
+def rule(**kwargs: Unpack[RuleDecoratorKwargs]) -> Callable[[F], F]:
+    """Handles decorator factories of the form `@rule(foo=..., bar=...)`
+    https://mypy.readthedocs.io/en/stable/generics.html#decorator-factories.
+
+    Note: This needs to be the first rule, otherwise MyPy goes nuts
+    """
+    ...
 
 
 @overload
-def rule(func: Callable[P, R]) -> Callable[P, Coroutine[Any, Any, R]]: ...
+def rule(_func: Callable[P, Coroutine[Any, Any, R]]) -> Callable[P, Coroutine[Any, Any, R]]:
+    """Handles bare @rule decorators on async functions.
+
+    Usage of Coroutine[...] (vs Awaitable[...]) is intentional, as `MultiGet`/`concurrently` use
+    coroutines directly.
+    """
+    ...
 
 
 @overload
-def rule(
-    *args, func: None = None, **kwargs: Any
-) -> Callable[[SyncRuleT | AsyncRuleT], AsyncRuleT]: ...
+def rule(_func: Callable[P, R]) -> Callable[P, Coroutine[Any, Any, R]]:
+    """Handles bare @rule decorators on non-async functions It's debatable whether we should even
+    have non-async @rule functions, but keeping this to not break the world for plugin authors.
+
+    Usage of Coroutine[...] (vs Awaitable[...]) is intentional, as `MultiGet`/`concurrently` use
+    coroutines directly.
+    """
+    ...
 
 
 def rule(*args, **kwargs):
```

Comment:
Technically you can spec this and it won't affect caller typing, but it will itself be internally type-checked.

Reply:
I thought about it, but my plan was to do that as part of the rules internals re-factor. I didn't want to obscure this PR any more than it needed to be (because the internal re-factor will touch like 150 lines).